array.go

```diff
@@ -913,3 +913,60 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+// Filter accepts a function that returns a boolean. The function is ran
+// over the array items. If the function returns true, the item passes
+// the filter and is added to the new array of filtered items. If false,
+// the item is skipped (or in other words filtered out). After iterating
+// through all the items of the original array, the new filtered array
+// is returned.
+//
+// If there are any failed assertions in the filtering function, the
+// item is omitted without causing chain failure.
+//
+// Example:
+//
+// object := make(map[string]interface{})
+// filteredObject := object.Filter(func(_ string, value *Value) bool {
+//	value.String().NotEmpty()                //fails on 1 and 2
+//  return value != "bar"                    //fails on "bar"
+// })
+//
+// filteredArray.Raw() == []interface{}{"foo"} //true
+
+func (a *Array) Filter(filter func(index int, value *Value) bool) *Array {
+	a.chain.enter("Filter()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return newArray(a.chain, nil)
+	}
+
+	if filter == nil {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected nil function argument"),
+			},
+		})
+	}
+
+	filteredArray := []interface{}{}
+
+	for index, element := range a.value {
+		chainFailed := false
+		valueChain := a.chain.clone()
+		valueChain.setFatal(false)
+
+		valueChain.setFailCallback(func() {
```

Comment:
Nitpick: it would be a bit more readble if you declare chainFailed right before setFailCallback call.